src/tools/kdb/external.c

```diff
@@ -0,0 +1,205 @@
+{
+	for (int i = 0; i < map->size; i++)
+	{
+		if (elektraStrCmp (map->array[i].commandName, key) == 0)
+		{
+			return map->array[i].binary;
+		}
+	}
+	return NULL;
+}
+
+void freeExternalBinaryMap (StringMap * map)
+{
+	for (int i = 0; i < map->size; i++)
+	{
+		if (map->array[i].commandName != NULL)
+		{
+			elektraFree (map->array[i].commandName);
+			map->array[i].commandName = NULL;
+		}
+		if (map->array[i].binary != NULL)
+		{
+			elektraFree (map->array[i].binary);
+			map->array[i].binary = NULL;
+		}
+	}
+	map->size = 0;
+}
+
+int tryLoadExternal (char * commandName, StringMap * map)
+{
+	char * execPathPtr = getenv ("KDB_EXEC_PATH");
+	bool found = false;
+	char path[1024] = { 0 };
+	struct stat buf;
+
+	if (execPathPtr)
+	{
+		char * execPath = strdup (execPathPtr);
+		char * token = strtok (execPath, ":");
+		while (token != NULL && !found)
+		{
+			snprintf (path, sizeof (path), "%s/%s", token, commandName);
+			found = stat (path, &buf) != -1;
+			token = strtok (NULL, ":");
+		}
+		elektraFree (execPath);
+	}
+
+	if (!found)
+	{
+		snprintf (path, sizeof (path), "%s/%s", BUILTIN_EXEC_FOLDER, commandName);
+		found = stat (path, &buf) != -1;
+	}
+
+	if (found)
+	{
+		insert (map, elektraStrDup (commandName), strdup (path));
+		return 0;
+	}
+	return 1;
+}
+
+int loadExternalSpec (KeySet * spec, StringMap * map, Key * errorKey)
+{
+	KDB * handle = kdbOpen (NULL, errorKey);
+	Key * baseKey = keyNew ("spec:" CLI_BASE_KEY, KEY_END);
+	KeySet * config = ksNew (0, KS_END);
+	if (kdbGet (handle, config, errorKey) == -1)
+	{
+		ELEKTRA_SET_VALIDATION_SEMANTIC_ERRORF (errorKey, "could not load '%s': %s", CLI_BASE_KEY, GET_ERR (baseKey));
+		keyDel (baseKey);
+		ksDel (config);
+		kdbClose (handle, errorKey);
+		return 1;
+	}
+	Key * cur = NULL;
+	KeySet * part = ksCut (config, baseKey);
+
+	for (elektraCursor it = 0; it < ksGetSize (part); ++it)
+	{
+		cur = ksAtCursor (part, it);
+		const Key * externalMeta = keyGetMeta (cur, "external");
+		const Key * externalBinary = keyGetMeta (cur, "bin");
+		bool isExternal = false;
+		if (externalBinary != NULL && externalMeta != NULL && elektraKeyToBoolean (externalMeta, &isExternal) && isExternal)
+		{ // add external spec and save path to binary
+			KeySet * externalCommandSpec = ksCut (part, cur);
+			insert (map, elektraStrDup (keyBaseName (cur)), elektraStrDup (keyString (externalBinary)));
+			ksAppend (spec, externalCommandSpec);
+			ksDel (externalCommandSpec);
+		}
+	}
+	ksDel (part);
+	ksDel (config);
+	kdbClose (handle, errorKey);
+	keyDel (baseKey);
+	return 0;
+}
+
+int runExternal (char * bin, char ** argv, Key * errorKey)
+{
+	// the external program should think it was called directly
+	argv[1] = bin;
+
+	pid_t pid;
+	int status = 0;
+	posix_spawnattr_t attr;
+	posix_spawn_file_actions_t fact;
+
+	posix_spawnattr_init (&attr);
+	posix_spawn_file_actions_init (&fact);
+
+	if (posix_spawn (&pid, bin, &fact, &attr, &(argv[1]), __environ) != 0)
```

Comment:
I like that we `fork` now and even avoid `fork` itself by using `posix_spawn`. Not sure, it's actually needed (a simple `execve` probably would be enough). However, is that really supported on Windows/MinGW? You're already switching to the Win32 API below, so you may as well do the entire process creating with the Win32 API.

Reply:
I wanted to keep as much code as possible shared. It was about waiting for it, so error printing/cleanup is still done by the CLI.